File: src/main/java/xyz/chunkstories/core/voxel/components/VoxelComponentSignText.java
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.voxel.components;

import xyz.chunkstories.api.voxel.components.VoxelComponent;
import xyz.chunkstories.api.world.cell.CellComponents;
import xyz.chunkstories.api.world.serialization.StreamSource;
import xyz.chunkstories.api.world.serialization.StreamTarget;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class VoxelComponentSignText extends VoxelComponent {
	public VoxelComponentSignText(CellComponents holder) {
		super(holder);
	}

	private String signText = "";
	public String getSignText() {
		return signText;
	}

	public void setSignText(String signText) {
		this.signText = signText;
		getHolder().getCell().refreshRepresentation();
	}

	@Override
	public void push(StreamTarget destinator, DataOutputStream dos) throws IOException {
		dos.writeUTF(signText);
	}

	@Override
	public void pull(StreamSource from, DataInputStream dis) throws IOException {
		signText = dis.readUTF();
	}
}

File: src/main/java/xyz/chunkstories/core/voxel/components/VoxelComponentSignText.kt
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.voxel.components

import xyz.chunkstories.api.content.json.Json
import xyz.chunkstories.api.content.json.asString
import xyz.chunkstories.api.voxel.components.VoxelComponent
import xyz.chunkstories.api.world.cell.CellComponents
import xyz.chunkstories.api.world.serialization.StreamSource
import xyz.chunkstories.api.world.serialization.StreamTarget

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

class VoxelComponentSignText(holder: CellComponents) : VoxelComponent(holder) {

	var signText = ""
		set(value) {
			field = value
			holder.cell.refreshRepresentation()
		}

	override fun serialize(): Json? = Json.Value.Text(signText)

	override fun deserialize(json: Json) {
		signText = json.asString ?: signText
	}
}
